consensus/core/src/synchronizer.rs

```diff
@@ -250,17 +409,111 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
     async fn process_fetched_blocks(
         serialized_blocks: Vec<Bytes>,
         peer_index: AuthorityIndex,
-        requested_block_refs: BTreeSet<BlockRef>,
+        requested_blocks_guard: BlocksGuard,
         core_dispatcher: Arc<D>,
         block_verifier: Arc<V>,
         context: Arc<Context>,
+        commands_sender: Sender<Command>,
     ) -> ConsensusResult<()> {
-        let mut verified_blocks = Vec::new();
+        // The maximum number of blocks that can be additionally fetched from the one requested - those
+        // are potentially missing ancestors.
+        const MAX_ADDITIONAL_BLOCKS: usize = 10;
+
+        // Split the number of blocks to the requested and additional ones
+        let at = serialized_blocks
+            .iter()
+            .position(|b| b.is_empty())
+            .unwrap_or(serialized_blocks.len());
+        let (serialized_blocks, mut ancestor_serialized_blocks) = serialized_blocks.split_at(at);
+
+        // remove the separator (empty bytes element) from the `ancestor_serialized_blocks` if not empty
+        if !ancestor_serialized_blocks.is_empty() {
+            ancestor_serialized_blocks = &ancestor_serialized_blocks[1..];
+        }
 
-        if serialized_blocks.len() > requested_block_refs.len() {
+        if serialized_blocks.len() > requested_blocks_guard.block_refs.len()
+            || ancestor_serialized_blocks.len() > MAX_ADDITIONAL_BLOCKS
+        {
             return Err(ConsensusError::TooManyFetchedBlocksReturned(peer_index));
         }
 
+        // Verify the requested blocks
+        let mut blocks = Self::verify_blocks(
```

Comment:
Is it verified that these blocks are subset of requested? If this verification exists, I think it will be unnecessary to have the separator too.

Reply:
It is verified yes. Having the separator though makes our live much easier here as we can first check the explicitly requested blocks with confidence, and then the additionally requested ones.